### ISSUE TITLE
switch aarch64 to ld.gold

### ISSUE THREAD

**Issue by nsabovic** [android-ndk/ndk#556]
When using NDK r16 Beta 2 (16.0.4414007-beta2), compiling with clang still uses ld.bfd on aarch64 (`aarch64-linux-android-4.9/prebuilt/darwin-x86_64/aarch64-linux-android/bin/ld` is `aarch64-linux-android-4.9/prebuilt/darwin-x86_64/aarch64-linux-android/bin/ld.bfd`).

This makes `-flto` fail, but only on ARM64, because the linker does not recognize LLVM bitcode files. LTO works just fine on arm and x86.

**Comment by DanAlbert** [comment]
Add `-flto=gold` to your ldflags.

We should probably change the default at this point though. It's been the default in the platform for quite a while now.

**Comment by nsabovic** [comment]
That worked, thanks so much!

I added `-Wl,-flto=gold` to `CMAKE_SHARED_LINKER_FLAGS_RELEASE` and it fixed the arm64 build.

FYI, adding that flag breaks the x86 build (which works fine without it) with a bunch of errors like this one:

    ~/Library/Android/sdk/ndk-bundle/toolchains/x86-4.9/prebuilt/darwin-x86_64/lib/gcc/i686-linux-android/4.9.x/../../../../i686-linux-android/bin/ld: error: ../../../../src/main/jniLibs/x86/Release/libippicv.a(jmp_icvippiDFTGetSize_C_32fc_as.s.o): relocation R_386_GOTOFF against preemptible symbol icv_ippJumpIndexForMergedLibs cannot be used when making a shared object

**Comment by DanAlbert** [comment]
Really? x86 already defaults to gold.

**Comment by DanAlbert** [comment]
Oh, I see a couple problems.

First, I gave you the wrong flag. It's `-fuse-ld=gold`. Second, it's `-fuse-ld=gold`, not `-Wl,-fuse-ld=gold`. `-Wl` flags are passed to the linker. If you've reached the linker, it's too late to choose which linker you want :) `-f` does very different things to the linker.

**Comment by nsabovic** [comment]
Now that worked perfectly! Thanks again!

If only [Clang's documentation](https://clang.llvm.org/docs/ClangCommandLineReference.html) did not look like a Siberian tundra—the flag is listed in there, but not what it does :(

Should I close this?

**Comment by DanAlbert** [comment]
No, I'm going to leave it open to track switching the default. Glad to hear that worked for you though.

**Comment by DanAlbert** [comment]
BTW, for a lot of Clang flags you can actually find pretty good documentation in the GCC docs. A fortunate side effect of Clang trying to maintain command line compatibility with GCC (though probably also a reason Clang's docs aren't as thorough as they ought to be).

**Comment by nsabovic** [comment]
Thanks, and you're right, it's well documented on the [gcc side](https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html#Link-Options).

**Comment by DanAlbert** [comment]
This doesn't actually block LTO any more. The latest Clang LTO works with bfd as well.

**Comment by rprichard** [comment]
> This doesn't actually block LTO any more. The latest Clang LTO works with bfd as well.

Is this true on Windows, too? I'm currently seeing an `LLVMgold.dll: error loading plugin` error with the bfd linker. To fix that, I have to enable Gold (using `-fuse-ld=gold.exe`, not `-fuse-ld=gold`).

**Comment by DanAlbert** [comment]
Ugh, that's frustrating. Unfortunately this is a no-go for r17 either way because of https://issuetracker.google.com/70838247.

**Comment by DanAlbert** [comment]
iirc I needed to fix gold to do the right thing when searching for LLVMgold (since Windows doesn't use rpath and by default only searches the executable's directory). Maybe we just need the same thing for bfd and can get that in time for r17 beta 2.

**Comment by rprichard** [comment]
The Clang driver is passing an absolute path to `LLVMgold.dll` to `ld.bfd.exe`.

I see that LTO works on 32-bit Windows, but not 64-bit Windows. I also see that `LLVMgold.dll` is provided both as part of the GCC toolchain and as part of the LLVM toolchain. For 32-bit Windows, the copy provided with the GCC toolchain is the wrong architecture -- it's a 64-bit binary.

**Comment by DanAlbert** [comment]
Still blocked by https://issuetracker.google.com/70838247.

**Comment by DanAlbert** [comment]
Given that we're now shipping lld, I'm guessing this won't ever happen and we'll just go straight to LLD.

**Comment by DanAlbert** [comment]
Closing in favor of lld.

**Comment by skamir786786** [comment]
Dishabling vdso32 solved my problem
